### PR TITLE
sunxi: add Orange Pi Zero 3 support

### PR DESCRIPTION
**Pull request by aiamadeus**
Key features:
 - Allwinner H618 SoC (Quad core Cortex-A53)
 - 1/1.5/2/4 GiB LPDDR4 DRAM
 - 1 USB 2.0 type C port (power + OTG)
 - 1 USB 2.0 host port
 - 1Gbps Ethernet port
 - Micro-HDMI port
 - MicroSD slot

Installation:
  Write the image to SD Card with dd.

### CODE REVIEW

**Comment by wigyori** [comment]
Hi - did you have a chance to test out the u-boot upgrade on non-H618 boards too i.e. A20 or H3 ?

**Comment by aiamadeus** [comment]
> Hi - did you have a chance to test out the u-boot upgrade on non-H618 boards too i.e. A20 or H3 ?

Sorry, I only have the h618 board now.

**Comment by wigyori** [comment]
> > Hi - did you have a chance to test out the u-boot upgrade on non-H618 boards too i.e. A20 or H3 ?
> 
> Sorry, I only have the h618 board now.

No worries, I will test out the build on some of them in the next few days.

**Comment by wigyori** [comment]
For the uboot upgrade:

Runtime-tested on: Olimex Olinuxino Micro (A20), Pine64 SoPine (A64).

Tested-by: Zoltan HERPAI <wigyori@uid0.hu>

**Comment by aiamadeus** [comment]
> For the uboot upgrade:
> 
> Runtime-tested on: Olimex Olinuxino Micro (A20), Pine64 SoPine (A64).

Thank you very much for your test. Can this pull request be merged?

File: target/linux/starfive/patches-6.1/0117-driver-mfd-axp20x-Add-support-for-AXP15060.patch
Comment (hauke) [inline target/linux/starfive/patches-6.1/0117-driver-mfd-axp20x-Add-support-for-AXP15060.patch:1]: Did you test this on a starfive?

Is this patch mostly a combination of the patches you backported?
Is the resulting code for the starfive board very similar after this change?

Comment (aiamadeus) [reply target/linux/starfive/patches-6.1/0117-driver-mfd-axp20x-Add-support-for-AXP15060.patch:1]: Sorry but I don't have the starfive board. Due to CI error I found that it conflicts with the patches in starfive.
After comparison, I think this patch just takes the upstream code and overwrites the driver on kernel 6.1.

Comment (hauke) [reply target/linux/starfive/patches-6.1/0117-driver-mfd-axp20x-Add-support-for-AXP15060.patch:1]: This PR does the following changes:
````
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x.c build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x.c 
--- backup/axp20x.c	2024-03-24 14:37:39.432418341 +0100
+++ build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x.c	2024-03-24 14:38:32.530401163 +0100
@@ -151,7 +151,6 @@
 
 /* AXP288 ranges are shared with the AXP803, as they cover the same range */
 static const struct regmap_range axp288_writeable_ranges[] = {
-	regmap_reg_range(AXP288_POWER_REASON, AXP288_POWER_REASON),
 	regmap_reg_range(AXP20X_DATACACHE(0), AXP20X_IRQ6_STATE),
 	regmap_reg_range(AXP20X_DCDC_MODE, AXP288_FG_TUNE5),
 };
@@ -696,7 +695,8 @@
 	.name			= "axp152_irq_chip",
 	.status_base		= AXP152_IRQ1_STATE,
 	.ack_base		= AXP152_IRQ1_STATE,
-	.unmask_base		= AXP152_IRQ1_EN,
+	.mask_base		= AXP152_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp152_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp152_regmap_irqs),
@@ -733,7 +733,8 @@
 	.name			= "axp20x_irq_chip",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp20x_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp20x_regmap_irqs),
@@ -745,7 +746,8 @@
 	.name			= "axp22x_irq_chip",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp22x_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp22x_regmap_irqs),
@@ -756,7 +758,8 @@
 	.name			= "axp288_irq_chip",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp288_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp288_regmap_irqs),
@@ -779,7 +782,8 @@
 	.name			= "axp803",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp803_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp803_regmap_irqs),
@@ -790,7 +794,8 @@
 	.name			= "axp806",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp806_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp806_regmap_irqs),
@@ -801,7 +806,8 @@
 	.name			= "axp809",
 	.status_base		= AXP20X_IRQ1_STATE,
 	.ack_base		= AXP20X_IRQ1_STATE,
-	.unmask_base		= AXP20X_IRQ1_EN,
+	.mask_base		= AXP20X_IRQ1_EN,
+	.mask_invert		= true,
 	.init_ack_masked	= true,
 	.irqs			= axp809_regmap_irqs,
 	.num_irqs		= ARRAY_SIZE(axp809_regmap_irqs),
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x.c build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x 
axp20x.c      axp20x-i2c.c  axp20x-rsb.c  
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x-i2c.c build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x-i2c.c 
--- backup/axp20x-i2c.c	2024-03-24 14:37:39.432418341 +0100
+++ build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x-i2c.c	2024-03-24 14:38:32.530401163 +0100
@@ -60,10 +60,12 @@
 #ifdef CONFIG_OF
 static const struct of_device_id axp20x_i2c_of_match[] = {
 	{ .compatible = "x-powers,axp152", .data = (void *)AXP152_ID },
+	{ .compatible = "x-powers,axp192", .data = (void *)AXP192_ID },
 	{ .compatible = "x-powers,axp202", .data = (void *)AXP202_ID },
 	{ .compatible = "x-powers,axp209", .data = (void *)AXP209_ID },
 	{ .compatible = "x-powers,axp221", .data = (void *)AXP221_ID },
 	{ .compatible = "x-powers,axp223", .data = (void *)AXP223_ID },
+	{ .compatible = "x-powers,axp313a", .data = (void *)AXP313A_ID },
 	{ .compatible = "x-powers,axp803", .data = (void *)AXP803_ID },
 	{ .compatible = "x-powers,axp806", .data = (void *)AXP806_ID },
 	{ .compatible = "x-powers,axp15060", .data = (void *)AXP15060_ID },
@@ -74,10 +76,12 @@
 
 static const struct i2c_device_id axp20x_i2c_id[] = {
 	{ "axp152", 0 },
+	{ "axp192", 0 },
 	{ "axp202", 0 },
 	{ "axp209", 0 },
 	{ "axp221", 0 },
 	{ "axp223", 0 },
+	{ "axp313a", 0 },
 	{ "axp803", 0 },
 	{ "axp806", 0 },
 	{ "axp15060", 0 },
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/
axp20x.c      axp20x.h      axp20x-i2c.c  
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x.h build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/include/^C
[hauke@hauke-p14s openwrt]$ cp^Cuild_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x-i2c.c build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/drivers/mfd/axp20x.c build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/include/linux/mfd/axp20x.h backup/
[hauke@hauke-p14s openwrt]$ diff -Naur backup/axp20x.h build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/include/linux/mfd/axp20x.h 
--- backup/axp20x.h	2024-03-24 14:37:39.432418341 +0100
+++ build_dir/target-riscv64_riscv64_musl/linux-starfive_generic/linux-6.1.82/include/linux/mfd/axp20x.h	2024-03-24 14:38:32.530401163 +0100
@@ -570,9 +570,8 @@
 	AXP152_IRQ_PEK_SHORT,
 	AXP152_IRQ_PEK_LONG,
 	AXP152_IRQ_TIMER,
-	/* out of bit order to make sure the press event is handled first */
-	AXP152_IRQ_PEK_FAL_EDGE,
 	AXP152_IRQ_PEK_RIS_EDGE,
+	AXP152_IRQ_PEK_FAL_EDGE,
 	AXP152_IRQ_GPIO3_INPUT,
 	AXP152_IRQ_GPIO2_INPUT,
 	AXP152_IRQ_GPIO1_INPUT,
@@ -647,9 +646,8 @@
 	AXP20X_IRQ_LOW_PWR_LVL1,
 	AXP20X_IRQ_LOW_PWR_LVL2,
 	AXP20X_IRQ_TIMER,
-	/* out of bit order to make sure the press event is handled first */
-	AXP20X_IRQ_PEK_FAL_EDGE,
 	AXP20X_IRQ_PEK_RIS_EDGE,
+	AXP20X_IRQ_PEK_FAL_EDGE,
 	AXP20X_IRQ_GPIO3_INPUT,
 	AXP20X_IRQ_GPIO2_INPUT,
 	AXP20X_IRQ_GPIO1_INPUT,
@@ -678,9 +676,8 @@
 	AXP22X_IRQ_LOW_PWR_LVL1,
 	AXP22X_IRQ_LOW_PWR_LVL2,
 	AXP22X_IRQ_TIMER,
-	/* out of bit order to make sure the press event is handled first */
-	AXP22X_IRQ_PEK_FAL_EDGE,
 	AXP22X_IRQ_PEK_RIS_EDGE,
+	AXP22X_IRQ_PEK_FAL_EDGE,
 	AXP22X_IRQ_GPIO1_INPUT,
 	AXP22X_IRQ_GPIO0_INPUT,
 };
@@ -758,9 +755,8 @@
 	AXP803_IRQ_LOW_PWR_LVL1,
 	AXP803_IRQ_LOW_PWR_LVL2,
 	AXP803_IRQ_TIMER,
-	/* out of bit order to make sure the press event is handled first */
-	AXP803_IRQ_PEK_FAL_EDGE,
 	AXP803_IRQ_PEK_RIS_EDGE,
+	AXP803_IRQ_PEK_FAL_EDGE,
 	AXP803_IRQ_PEK_SHORT,
 	AXP803_IRQ_PEK_LONG,
 	AXP803_IRQ_PEK_OVER_OFF,
@@ -811,9 +807,8 @@
 	AXP809_IRQ_LOW_PWR_LVL1,
 	AXP809_IRQ_LOW_PWR_LVL2,
 	AXP809_IRQ_TIMER,
-	/* out of bit order to make sure the press event is handled first */
-	AXP809_IRQ_PEK_FAL_EDGE,
 	AXP809_IRQ_PEK_RIS_EDGE,
+	AXP809_IRQ_PEK_FAL_EDGE,
 	AXP809_IRQ_PEK_SHORT,
 	AXP809_IRQ_PEK_LONG,
 	AXP809_IRQ_PEK_OVER_OFF,
@@ -904,4 +899,4 @@
  */
 void axp20x_device_remove(struct axp20x_dev *axp20x);
 
-#endif /* __LINUX_MFD_AXP20X_H */
\ No newline at end of file
+#endif /* __LINUX_MFD_AXP20X_H */
[hauke@hauke-p14s openwrt]$ 
````

Please backport this patches too:
https://github.com/torvalds/linux/commit/8781ba7f45695af3ab8e8d1b55a31f527c9201a3
https://github.com/torvalds/linux/commit/2405fbfb384ef39e9560d76d3f6e4c90519f90aa

The .unmask_base part is just some styling, see https://patches.linaro.org/project/alsa-devel/patch/20220620200644.1961936-29-aidanmacdonald.0x0@gmail.com/

Comment (aiamadeus) [reply target/linux/starfive/patches-6.1/0117-driver-mfd-axp20x-Add-support-for-AXP15060.patch:1]: > Please backport this patches too:
> https://github.com/torvalds/linux/commit/8781ba7f45695af3ab8e8d1b55a31f527c9201a3
> https://github.com/torvalds/linux/commit/2405fbfb384ef39e9560d76d3f6e4c90519f90aa

Thanks for your suggestion, I will backport it later.

**Pull request merged**